File: targets/android-widgets/kotlin/Snapshot.kt
package com.owaiskhan.monthlyplanning.widgets

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * Mirrors `WidgetSnapshot` in src/widgets/snapshot.ts, same as the Swift
 * `PlannerSnapshot`. Parsed with org.json rather than a serialization library
 * so the widgets pull in no extra dependencies.
 */
data class Habit(val id: String, val name: String)

data class Goal(val text: String, val done: Boolean)

/**
 * A task with a deadline. `due` is the moment itself in epoch milliseconds, not
 * a distance to it — the widget works out how far off it is against its own
 * clock, so the countdown stays true between snapshots.
 */
data class DeadlineTask(val text: String, val due: Long)

data class PlannerSnapshot(
  /**
   * The app's appearance setting — "dark" or "light". Null for snapshots
   * written before the widgets followed the app, and for the picker preview;
   * both fall back to the system setting.
   */
  val theme: String?,
  val year: Int,
  val month: Int,
  val monthName: String,
  val daysInMonth: Int,
  val today: Int?,
  val habits: List<Habit>,
  val grid: Map<String, String>,
  val monthDone: Int,
  val monthMissed: Int,
  val monthTotal: Int,
  val monthPct: Int,
  val currentStreak: Int,
  val bestStreak: Int,
  val perfectStreak: Int,
  val goals: List<Goal>,
  val yearDone: List<List<Int>>,
  val yearMissed: List<List<Int>>,
  val yearHabitCounts: List<Int>,
  val yearTotal: Int,
  val quotes: List<String>,
  /** unfinished deadlines, soonest first — overdue ones included, and first */
  val tasks: List<DeadlineTask>,
) {
  /** 0 pending, 1 done, 2 missed */
  fun state(day: Int, habitIndex: Int): Int {
    val row = grid[day.toString()] ?: return 0
    if (habitIndex < 0 || habitIndex >= row.length) return 0
    return row[habitIndex].digitToIntOrNull() ?: 0
  }

  fun doneCount(day: Int): Int = grid[day.toString()]?.count { it == '1' } ?: 0

  /** Same day-of-year rotation as quoteForDate in src/quotes.ts. */
  fun quote(dayOfYear: Int): String? =
    if (quotes.isEmpty()) null else quotes[dayOfYear % quotes.size]

  companion object {
    fun load(context: Context): PlannerSnapshot? {
      val json = context
        .getSharedPreferences("widget_snapshot", Context.MODE_PRIVATE)
        .getString("widgetSnapshot", null) ?: return null
      return runCatching { parse(JSONObject(json)) }.getOrNull()
    }

    private fun intMatrix(array: JSONArray?): List<List<Int>> {
      if (array == null) return emptyList()
      return (0 until array.length()).map { i ->
        val row = array.optJSONArray(i) ?: JSONArray()
        (0 until row.length()).map { row.optInt(it) }
      }
    }

    private fun parse(o: JSONObject): PlannerSnapshot {
      val habits = o.optJSONArray("habits").let { arr ->
        (0 until (arr?.length() ?: 0)).map {
          val h = arr!!.getJSONObject(it)
          Habit(h.optString("id"), h.optString("name"))
        }
      }
      val goals = o.optJSONArray("goals").let { arr ->
        (0 until (arr?.length() ?: 0)).map {
          val g = arr!!.getJSONObject(it)
          Goal(g.optString("text"), g.optBoolean("done"))
        }
      }
      val grid = mutableMapOf<String, String>()
      o.optJSONObject("grid")?.let { g ->
        g.keys().forEach { key -> grid[key] = g.optString(key) }
      }
      val quotes = o.optJSONArray("quotes").let { arr ->
        (0 until (arr?.length() ?: 0)).map { arr!!.optString(it) }
      }
      // absent from snapshots written before deadlines existed, which must
      // still parse rather than taking every other widget down with them
      val tasks = o.optJSONArray("tasks").let { arr ->
        (0 until (arr?.length() ?: 0)).map {
          val t = arr!!.getJSONObject(it)
          DeadlineTask(t.optString("text"), t.optLong("due"))
        }
      }.sortedBy { it.due }
      return PlannerSnapshot(
        theme = if (o.isNull("theme")) null else o.optString("theme").ifEmpty { null },
        year = o.optInt("year"),
        month = o.optInt("month"),
        monthName = o.optString("monthName"),
        daysInMonth = o.optInt("daysInMonth", 30),
        today = if (o.isNull("today")) null else o.optInt("today"),
        habits = habits,
        grid = grid,
        monthDone = o.optInt("monthDone"),
        monthMissed = o.optInt("monthMissed"),
        monthTotal = o.optInt("monthTotal"),
        monthPct = o.optInt("monthPct"),
        currentStreak = o.optInt("currentStreak"),
        bestStreak = o.optInt("bestStreak"),
        perfectStreak = o.optInt("perfectStreak"),
        goals = goals,
        yearDone = intMatrix(o.optJSONArray("yearDone")),
        yearMissed = intMatrix(o.optJSONArray("yearMissed")),
        yearHabitCounts = o.optJSONArray("yearHabitCounts").let { arr ->
          (0 until (arr?.length() ?: 0)).map { arr!!.optInt(it) }
        },
        yearTotal = o.optInt("yearTotal"),
        quotes = quotes,
        tasks = tasks,
      )
    }

    /** Shown in the widget picker and before the app has ever written a snapshot. */
    val placeholder = PlannerSnapshot(
      // the picker preview has no app setting to follow yet
      theme = null,
      year = 2026,
      month = 7,
      monthName = "AUGUST",
      daysInMonth = 31,
      today = 15,
      habits = listOf(
        Habit("0", "Read 20 pages"),
        Habit("1", "Gym"),
        Habit("2", "No sugar"),
        Habit("3", "Deep work"),
      ),
      grid = (1..15).associate { it.toString() to listOf("1110", "1101", "1111", "1011", "1210")[it % 5] },
      monthDone = 52,
      monthMissed = 6,
      monthTotal = 124,
      monthPct = 42,
      currentStreak = 9,
      bestStreak = 21,
      perfectStreak = 3,
      goals = listOf(Goal("Ship v1", true), Goal("Run 50k", false), Goal("Read 3 books", false)),
      yearDone = (0 until 12).map { m -> (0 until 30).map { if (m <= 7) (it % 4) else 0 } },
      yearMissed = (0 until 12).map { (0 until 30).map { 0 } },
      yearHabitCounts = List(12) { 4 },
      yearTotal = 318,
      quotes = listOf("Discipline is choosing between what you want now and what you want most."),
      tasks = listOf(
        DeadlineTask("Send the tax return", System.currentTimeMillis() + 5 * 3_600_000L),
        DeadlineTask("Book the flights", System.currentTimeMillis() + 2 * 86_400_000L),
        DeadlineTask("Finish the draft", System.currentTimeMillis() + 6 * 86_400_000L),
      ),
    )
  }
}

File: targets/widgets/Snapshot.swift
import Foundation
import WidgetKit

/// Mirrors `WidgetSnapshot` in src/widgets/snapshot.ts. Both sides must change together.
struct PlannerSnapshot: Codable {
  struct Habit: Codable {
    let id: String
    let name: String
  }

  struct Goal: Codable {
    let text: String
    let done: Bool
  }

  struct Task: Codable {
    let text: String
    /// The deadline, epoch milliseconds — the app's clock, not a countdown.
    /// Storing the moment rather than the distance to it is what lets the
    /// widget stay accurate between snapshots.
    let due: Double

    var date: Date { Date(timeIntervalSince1970: due / 1000) }
  }

  let updatedAt: Double
  /// The app's appearance setting — "dark" or "light". Optional because
  /// snapshots written before the widgets followed the app have no such key,
  /// and a missing one falls back to the system scheme rather than failing.
  let theme: String?
  let year: Int
  /// 0-based, matching the app
  let month: Int
  let monthName: String
  let daysInMonth: Int
  let today: Int?
  let habits: [Habit]
  /// day → one character per habit: "0" pending, "1" done, "2" missed
  let grid: [String: String]
  let monthDone: Int
  let monthMissed: Int
  let monthTotal: Int
  let monthPct: Int
  let currentStreak: Int
  let bestStreak: Int
  let perfectStreak: Int
  let goals: [Goal]
  let yearDone: [[Int]]
  let yearMissed: [[Int]]
  let yearHabitCounts: [Int]
  let yearTotal: Int
  /// Decoded leniently — snapshots written before the quote widget existed
  /// have no `quotes` key, and a missing one shouldn't fail the whole decode.
  let quotes: [String]?
  /// Likewise optional: a snapshot written before deadlines existed carries no
  /// `tasks` key, and that must not fail the decode for every other widget.
  let tasks: [Task]?

  /// Unfinished deadlines, soonest first — overdue ones included, and first in
  /// the list, since a deadline that has gone by is the one worth showing.
  var deadlines: [Task] {
    (tasks ?? []).sorted { $0.due < $1.due }
  }

  /// Same day-of-year rotation as `quoteForDate` in src/quotes.ts, so the app
  /// and the widget always show the same quote on a given day.
  func quote(on date: Date) -> String? {
    guard let quotes, !quotes.isEmpty else { return nil }
    let cal = Calendar.current
    guard let dayOfYear = cal.ordinality(of: .day, in: .year, for: date) else { return nil }
    return quotes[dayOfYear % quotes.count]
  }

  /// 0 pending, 1 done, 2 missed
  func state(day: Int, habitIndex: Int) -> Int {
    guard let row = grid[String(day)] else { return 0 }
    let chars = Array(row)
    guard habitIndex >= 0, habitIndex < chars.count else { return 0 }
    return Int(String(chars[habitIndex])) ?? 0
  }

  func doneCount(day: Int) -> Int {
    guard let row = grid[String(day)] else { return 0 }
    return row.filter { $0 == "1" }.count
  }

  func missedCount(day: Int) -> Int {
    guard let row = grid[String(day)] else { return 0 }
    return row.filter { $0 == "2" }.count
  }

  /// Shown before the app has ever written a snapshot, and in the widget gallery.
  static let placeholder: PlannerSnapshot = {
    let now = Date()
    let cal = Calendar.current
    let year = cal.component(.year, from: now)
    let month = cal.component(.month, from: now) - 1
    let day = cal.component(.day, from: now)
    let range = cal.range(of: .day, in: .month, for: now)?.count ?? 30
    let names = ["JANUARY", "FEBRUARY", "MARCH", "APRIL", "MAY", "JUNE",
                 "JULY", "AUGUST", "SEPTEMBER", "OCTOBER", "NOVEMBER", "DECEMBER"]

    var grid: [String: String] = [:]
    for d in 1...max(day, 1) {
      // a plausible-looking sample so the gallery preview isn't empty
      let pattern = ["1110", "1101", "1111", "1011", "1210"][d % 5]
      grid[String(d)] = pattern
    }

    var yearDone: [[Int]] = []
    var yearMissed: [[Int]] = []
    for m in 0..<12 {
      var comps = DateComponents()
      comps.year = year
      comps.month = m + 1
      let len = cal.range(of: .day, in: .month, for: cal.date(from: comps) ?? now)?.count ?? 30
      yearDone.append((0..<len).map { m <= month ? ($0 % 4 == 0 ? 0 : ($0 % 3) + 1) : 0 })
      yearMissed.append(Array(repeating: 0, count: len))
    }

    return PlannerSnapshot(
      updatedAt: now.timeIntervalSince1970 * 1000,
      // the gallery preview has no app setting to follow yet
      theme: nil,
      year: year,
      month: month,
      monthName: names[min(max(month, 0), 11)],
      daysInMonth: range,
      today: day,
      habits: [
        .init(id: "0", name: "Read 20 pages"),
        .init(id: "1", name: "Gym"),
        .init(id: "2", name: "No sugar"),
        .init(id: "3", name: "Deep work"),
      ],
      grid: grid,
      monthDone: 52,
      monthMissed: 6,
      monthTotal: 124,
      monthPct: 42,
      currentStreak: 9,
      bestStreak: 21,
      perfectStreak: 3,
      goals: [
        .init(text: "Ship v1", done: true),
        .init(text: "Run 50k", done: false),
        .init(text: "Read 3 books", done: false),
      ],
      yearDone: yearDone,
      yearMissed: yearMissed,
      yearHabitCounts: Array(repeating: 4, count: 12),
      yearTotal: 318,
      quotes: [
        "Discipline is choosing between what you want now and what you want most.",
        "We are what we repeatedly do. Excellence, then, is not an act, but a habit.",
        "You don't have to be extreme, just consistent.",
      ],
      tasks: [
        .init(text: "Send the tax return", due: now.addingTimeInterval(5 * 3600).timeIntervalSince1970 * 1000),
        .init(text: "Book the flights", due: now.addingTimeInterval(2 * 86400).timeIntervalSince1970 * 1000),
        .init(text: "Finish the draft", due: now.addingTimeInterval(6 * 86400).timeIntervalSince1970 * 1000),
      ]
    )
  }()
}

enum SnapshotStore {
  /// Must match app.json, expo-target.config.js and WidgetBridgeModule.swift.
  static let appGroup = "group.com.owaiskhan.monthlyplanning"
  static let key = "widgetSnapshot"

  static func load() -> PlannerSnapshot? {
    guard
      let defaults = UserDefaults(suiteName: appGroup),
      let json = defaults.string(forKey: key),
      let data = json.data(using: .utf8)
    else { return nil }
    return try? JSONDecoder().decode(PlannerSnapshot.self, from: data)
  }
}

struct SnapshotEntry: TimelineEntry {
  let date: Date
  let snapshot: PlannerSnapshot
  /// true when no snapshot has been written yet, so views can nudge the user
  let isPlaceholder: Bool
}

/// One provider serves every widget — they all render from the same snapshot.
struct SnapshotProvider: TimelineProvider {
  func placeholder(in context: Context) -> SnapshotEntry {
    SnapshotEntry(date: Date(), snapshot: .placeholder, isPlaceholder: true)
  }

  func getSnapshot(in context: Context, completion: @escaping (SnapshotEntry) -> Void) {
    let loaded = SnapshotStore.load()
    completion(
      SnapshotEntry(
        date: Date(),
        snapshot: loaded ?? .placeholder,
        isPlaceholder: loaded == nil
      )
    )
  }

  func getTimeline(in context: Context, completion: @escaping (Timeline<SnapshotEntry>) -> Void) {
    let loaded = SnapshotStore.load()
    let snapshot = loaded ?? .placeholder
    let now = Date()

    // The app reloads timelines whenever the data changes, so a timeline only
    // has to cover what moves on its own: "today" rolling over at midnight, and
    // every countdown stepping down. Those are entries, not reloads — WidgetKit
    // budgets how often an extension may be woken, but not how many entries it
    // hands back when it is.
    let midnight =
      Calendar.current.nextDate(
        after: now,
        matching: DateComponents(hour: 0, minute: 1),
        matchingPolicy: .nextTime
      ) ?? now.addingTimeInterval(3600)

    var dates = [now]
    dates += Deadline.labelChanges(after: now, in: snapshot, limit: 28)
    dates.append(midnight)
    dates = Array(Set(dates)).sorted()

    let entries = dates.map {
      SnapshotEntry(date: $0, snapshot: snapshot, isPlaceholder: loaded == nil)
    }
    completion(Timeline(entries: entries, policy: .after(dates.last ?? midnight)))
  }
}
